### ISSUE TITLE
AppState.currentState sometimes has value "uninitialized" indefinitely

### ISSUE THREAD

**Issue by rigdern** [Microsoft/react-native-windows#955]
I'm using the `currentState` property on React Native's `AppState` JavaScript module. Sometimes, I'm seeing that `AppState.currentState` has the value `"uninitialized"` indefinitely. I debugged this a bit but I don't fully understand the root cause.

Here's what I'm seeing happen in the react-native-windows C# `AppStateModule` in this order:
  - [`getCurrentAppState`](https://github.com/ReactWindows/react-native-windows/blob/393334bd718b5bb01cdc0e9c7a40d5fa3a23ed82/ReactWindows/ReactNative/Modules/AppState/AppStateModule.cs#L76) is called and resolves with `"uninitialized"`
  - [`OnResume`](https://github.com/ReactWindows/react-native-windows/blob/393334bd718b5bb01cdc0e9c7a40d5fa3a23ed82/ReactWindows/ReactNative/Modules/AppState/AppStateModule.cs#L57)
  - [`SendAppStateChangeEvent`](https://github.com/ReactWindows/react-native-windows/blob/393334bd718b5bb01cdc0e9c7a40d5fa3a23ed82/ReactWindows/ReactNative/Modules/AppState/AppStateModule.cs#L89) with `_appState` equal to `"active"`.

Here are the responses I'm seeing React Native's `AppState` JavaScript module get. Notice that the responses are received in the opposite order that react-native-windows queued them in:
  - [`appStateDidChange`](https://github.com/facebook/react-native/blob/v0.34.0-rc.0/Libraries/AppState/AppState.js#L95-L97) fires with value `"active"`
  - [`getCurrentAppState`](https://github.com/facebook/react-native/blob/v0.34.0-rc.0/Libraries/AppState/AppState.js#L104-L106) resolves with value `"uninitialized"`

This makes `AppState.currentState`'s final value `"uninitialized"`

In summary, it seems that the JavaScript `AppState` module isn't receiving the data in the order that the C# `AppStateModule` queued it. Is it expected that when C# resolves a callback and fires an event, the order these actions are received in could be different than the order they were queued in?

I reproed this with react-native-windows 393334bd718b5bb01cdc0e9c7a40d5fa3a23ed82 and React Native 0.34.0-rc.0.

**Comment by rigdern** [comment]
TLDR I think the C# `AppStateModule` should always access `_appState` on the same thread. And from that thread, it should also add operations to the JavaScript queue.

I now understand one cause for the behavior described in the bug report. The C# `AppStateModule` is currently running code on multiple threads simultaneously:
  - [`getCurrentAppState`](https://github.com/ReactWindows/react-native-windows/blob/393334bd718b5bb01cdc0e9c7a40d5fa3a23ed82/ReactWindows/ReactNative/Modules/AppState/AppStateModule.cs#L76) runs on one thread (probably the thread that dispatches calls to native modules on behalf of JavaScript)
  - [`OnResume`](https://github.com/ReactWindows/react-native-windows/blob/393334bd718b5bb01cdc0e9c7a40d5fa3a23ed82/ReactWindows/ReactNative/Modules/AppState/AppStateModule.cs#L57) runs on another thread (probably the UI thread)

In my description from my initial bug report, although I saw the C# `getCurrentAppState` method starting before the C# `OnResume` method started, I suspect that the CPU switched the thread it was running and `OnResume` ended up finishing and queuing its event before `getCurrentAppState` had a chance to finish and queue its callback. This would be an explanation for why I thought C# was queuing stuff in a different order than JavaScript was receiving it.

Anyway, it appears that the C# `AppStateModule` runs on multiple threads simultaneously but is not thread safe. For example, the `_appState` instance variable is accessed from multiple threads.

I thought a bit about the right fix for the C# `AppStateModule`. We need a pattern for the general problem of binding a JavaScript value to a changing native value. It's made complicated by all communication between JavaScript and C# being asynchronous. As an example, I'll say we want to keep JavaScript's `currentState` variable in sync with C#'s `_appState` variable.

First of all, C# needs to fire an event JavaScript can consume so it can update `currentState` whenever `_appState` changes. The ordering of the events is important. The most recent event JavaScript receives needs to contain the most recent value.

But the event isn't enough because JavaScript needs to know the value of the C# variable before it changes and an event fires. For this, C# can expose an async function so JavaScript can get the value on demand (e.g. `getCurrentAppState`). Again, ordering is important. The most recent **event** or **callback** JavaScript receives needs to contain the most recent value.

I think the easiest way to achieve this is for `AppStateModule` to always access `_appState` on the same thread. It should also use this thread to add events and callback resolutions to the JavaScript queue.

**Comment by rozele** [comment]
@rigdern it sounds like this is not actually a bug, but a symptom of the react-native threading model. If I'm not mistaken, the exact same behavior could occur on Android, where JavaScript triggers a call to `getCurrentState` simultaneously with when OnResume is called by the UIThread. I imagine this can all be handled from JavaScript by ensuring that the appStateDidChange event takes priority over any calls to `getCurrentState`. Closing for now, please re-open if you feel this is still an issue.

**Comment by rigdern** [comment]
@rozele Okay, we can fix this in the JS code in React Native by ignoring the result of `getCurrentAppState` if it arrives after the event has fired.

I don't think the instance variable `_appState` in the `AppState` module is used in a thread-safe way. It's read from multiple threads. I filed #1300 to track this.
